Updated LandingPage->PageBuilder migration documentation (#984)
ezsystems/developer-documentation: docs/updating/4_update_2.2.md

The command will migrate Landing Pages created in eZ Platform 1.x, 2.0 and 2.1 to new Pages.
The operation is transactional and will roll back in case of errors.

> In 2.2 Page Builder does not offer all blocks that Landing Page editor did. The removed blocks include Schedule and Form blocks.

Author: ```suggestion
In 2.2 Page Builder does not offer all blocks that Landing Page editor did. The removed blocks include Keyword, Schedule and Form blocks.
```

> The Page Builder in 2.2 does not offer all blocks that Landing Page editor did. The removed blocks include Schedule and Form blocks.

Reviewer: These exists on 2.5 right? So if possible to change this to approach migration on 2.5 version it won't need to be mentioned or? (or do we need to improve migration for these blocks maybe?)

Author: They exist on 2.5, yes. But it is still not possible to migrate to them if running migration script on 2.5. I also found that Tag block has not been added to pagebuilder.. So I'll add info about this one too

> The Page Builder in 2.2 does not offer all blocks that Landing Page editor did. The removed blocks include Schedule and Form blocks.

Reviewer 2: ```suggestion
In 2.2 Page Builder does not offer all blocks that Landing Page editor did. The removed blocks include Schedule and Form blocks.
```

> The Places Page Builder block has been removed from the clean installation and will only be available in the demo out of the box. If you had been using this block in your site, re-apply its configuration based on the [demo](https://github.com/ezsystems/ezplatform-ee-demo/blob/v2.2.2/app/config/blocks.yml)

Reviewer 2: ```suggestion
The Places block has been removed from the clean installation and will only be available in the demo out of the box. If you had been using this block in your site, re-apply its configuration based on the [demo](https://github.com/ezsystems/ezplatform-ee-demo/blob/v2.2.2/app/config/blocks.yml)
```

> Later versions of Page Builder come with a Content Scheduler block, but migration of Schedule blocks to Content Scheduler blocks is not supported. 

Author: ```suggestion
Later versions of Page Builder come with a Content Scheduler block and new Form block, but migration of Schedule blocks to Content Scheduler blocks and Form blocks are not supported. 
```

> Later versions of pagebuilder has support for the "Content Scheduler" block, but migration of Schedule blocks to this "Content Scheduler" block is not supported. 

Reviewer 2: ```suggestion
Later versions of Page Builder come with a Content Scheduler block, but migration of Schedule blocks to Content Scheduler blocks is not supported. 
```

After the migration is finished, you need to clear cache.

### Migrate layouts

> The `ez_block:renderBlockAction` controller used in layout templates has been replaced by `EzPlatformPageFieldTypeBundle:Block:render`. This controller has to additional parameters, `locationId` and `languageCode`. Only `languageCode` is required.

Reviewer 2: ```suggestion
The `ez_block:renderBlockAction` controller used in layout templates has been replaced by `EzPlatformPageFieldTypeBundle:Block:render`. This controller has two additional parameters, `locationId` and `languageCode`. Only `languageCode` is required.
```

> Also, the html class `data-studio-zone` has been replaced with `data-ez-zone-id`

Reviewer 2: ```suggestion
Also, the HTML class `data-studio-zone` has been replaced with `data-ez-zone-id`
```

See [documentation](../guide/page_rendering.md#layout-template) for an example on usage of the new controller.

### Migrate custom blocks

> Landingpage blocks (2.1 and earlier) was defined using a class implementing `EzSystems\LandingPageFieldTypeBundle\FieldType\LandingPage\Model\AbstractBlockType`. 

Reviewer 2: ```suggestion
Landing Page blocks (from 2.1 and earlier) were defined using a class implementing `EzSystems\LandingPageFieldTypeBundle\FieldType\LandingPage\Model\AbstractBlockType`. 
```

> In pagebuilder (2.2), this interface is no longer present. Instead the logic of your block must be implemented in a [Listener](../guide/extending/extending_page.md#block-rendering-events).

Reviewer 2: ```suggestion
In Page Builder (from 2.2 onwards), this interface is no longer present. Instead the logic of your block must be implemented in a [Listener](../guide/extending/extending_page.md#block-rendering-events).
```

Typically, what you previously would do in `getTemplateParameters()`, you'll now do in the `onBlockPreRender()` event handler.

The definition of block parameters has to be moved from `createBlockDefinition()` to the [YAML configuration](../guide/extending/extending_page.md#creating-page-blocks) for your custom blocks.

> More information about how custom blocks are implemented in Pagebuilder, please have a look at the [documentation](../guide/extending/extending_page.md)

Reviewer 2: ```suggestion
For more information about how custom blocks are implemented in Pagebuilder, have a look at [Creating custom Page blocks](../extending/extending_page.md)
```

> For the migration of blocks from landingpage you Pagebuilder, you'll need to provide a converter for attributes of custom blocks. For simple blocks you can use `\EzSystems\EzPlatformPageMigration\Converter\AttributeConverter\DefaultConverter`.

Reviewer 2: ```suggestion
For the migration of blocks from Landing Page to Page Builder, you'll need to provide a converter for attributes of custom blocks. For simple blocks you can use `\EzSystems\EzPlatformPageMigration\Converter\AttributeConverter\DefaultConverter`.
```

Custom converters must implement the `\EzSystems\EzPlatformPageMigration\Converter\AttributeConverter\ConverterInterface` interface.
`convert()` will parse XML `\DOMNode $node` and return an array of `\EzSystems\EzPlatformPageFieldType\FieldType\LandingPage\Model\Attribute` objects.

Below is an example of a simple converter for a custom block:

``` yaml
    app.block.foobar.converter:
        class: EzSystems\EzPlatformPageMigration\Converter\AttributeConverter\DefaultConverter
        tags:
            - { name: ezplatform.fieldtype.ezlandingpage.migration.attribute.converter, block_type: foobar }
```

> Notice the service tag that must be used for attribute converters.

Reviewer 2: ```suggestion
Notice service tag `ezplatform.fieldtype.ezlandingpage.migration.attribute.converter` that must be used for attribute converters.
```

This converter is only needed when running the `ezplatform:page:migrate` script and can be removed once that has completed.

#### Example on how to migrate layouts and blocks to the new Page Builder

> In the `upgraded_to_2.2` branch in this [git repo](https://github.com/ezsystems/ezplatform-ee-beginner-tutorial/tree/upgraded_to_2.2) you may see an example on how to migrate your source code in order to make a layout and a block work with the new Page Builder.

Reviewer 2: How about adding the `src/AppBundle/Resources/config/blocks.yml` and `/Event/Listener/RandomBlockListener.php` from the example here instead of linking? To make the page less cluttered, we could place it in an expanding code box.

Author: I added the code in expanding code box. Please have a look
